design_documents/prepare_cheatcode.md

```diff
@@ -0,0 +1,201 @@
+will yield different addresses.
+
+### `precalculate_address` Cheatcode
+
+Introduce a cheatcode with the signature:
+
+```cairo
+fn precalculate_address(prepared_contract: PreparedContract) -> ContractAddress;
+```
+
+That will use the same method of calculating the contract address as `deploy` uses, utilizing the internal counter.
+This way the user will have an ability to know the address of the contract that will be deployed, and the current
+deployment flow will remain unchanged.
+
+### Known Problems With This Solution
+
+For the address returned by `precalculate_address` to match the address from `deploy`, the user will have to
+call `precalculate_address` immediately before the deployment or at least before any other calls to `deploy` as the
+internal counter will then be incremented.
+
+This could be remedied by having separate counters for all `class_hashe`es, but it will still remain a limiting factor.
+
+## Example Usage
+
+```cairo
+mod HelloStarknet {
+    // ...
+    
+    #[constructor]
+    fn constructor(ref self: ContractState) {
+        let timestamp = starknet::get_block_timestamp();
+        self.create_time.write(timestamp);
+    }
+}
+
+#[test]
+fn call_and_invoke() {
+    // Declare the contract
+    let class_hash = declare('HelloStarknet');
+    
+    // Prepare contract for deployment
+    let prepared = PreparedContract { class_hash: class_hash, constructor_calldata: @ArrayTrait::new() };
+    
+    // Precalculate the address
+    let contract_address = precalulucate_address(prepared);
+    
+    // Warp the address
+    start_warp(contract_address, 1234);
+    
+    // Deploy with warped constructor
+    let contract_address = deploy(prepared).unwrap();
+    let dispatcher = IHelloStarknetDispatcher { contract_address };
+}
+```
+
+In case the user does not need to apply cheatcodes to the constructor, the deployment flow remains as before, without
+any steps introduced.
+
+```cairo
+#[test]
+fn call_and_invoke() {
+    // Declare the contract
+    let class_hash = declare('HelloStarknet');
+    
+    // Prepare contract for deployment
+    let prepared = PreparedContract { class_hash: class_hash, constructor_calldata: @ArrayTrait::new() };
+        
+    // Deploy
+    let contract_address = deploy(prepared).unwrap();
+    let dispatcher = IHelloStarknetDispatcher { contract_address };
+}
+```
+
+## Proposed Solution 2
+
+Change the current deployment flow, so it can better facilitate precalculating of contract addresses.
+
+### `declare` Cheatcode
+
+Change the `declare` cheatcode signature to this:
+
+```cairo
+trait DeclaredContractTrait {
+    fn precalculate_address(constructor_calldata: @Array::<felt252>) -> ContractAddress;
+    fn deploy(constructor_calldata: @Array::<felt252>) -> -> Result::<ContractAddress, RevertedTransaction>;
+}
+
+fn declare(contract: felt252) -> DeclaredContractTrait;
```

Comment:
Is it possible to return statically dispatched objects that implement traits in Cairo?
If not, it would be more like:

```cairo
struct DeclaredContract {
    class_hash: ClassHash,
    // not sure if more fields are needed
}

impl DeclaredContractTraitImpl of DeclaredContractTrait {
    // use `cheatcode` function from starknet to implement the trait
}

fn declare(contract: felt252) -> DeclaredContract;
```

Reply:
I think it's not possible, will change it to something similar to suggested.